nomos-utils/src/overwatch_recovery.rs

```diff
@@ -0,0 +1,286 @@
+where
+    Backend: RecoveryBackend + FromSettings<Settings = Settings> + Debug + Clone,
+    Backend::State: Default + Serialize + DeserializeOwned + Debug + Clone,
+{
+    type Settings = Settings;
+    type Error = MyError;
+
+    fn from_settings(settings: &Self::Settings) -> Result<Self, Self::Error> {
+        let state = Backend::from_settings(settings)
+            .load_state()
+            .unwrap_or_default();
+        Ok(Self {
+            state,
+            settings: Default::default(),
+            backend: Default::default(),
+        })
+    }
+}
+
+trait RecoveryBackend {
+    type State;
+    fn load_state(&self) -> _Result<Self::State>;
+    fn save_state(&self, state: &Self::State) -> _Result<()>;
+}
+
+#[derive(Debug, Clone)]
+struct RecoveryOperator<Backend, Settings>
+where
+    Backend: RecoveryBackend + Clone + Debug,
+    Backend::State: Clone + Debug,
+    Settings: Clone + Debug,
+{
+    recovery_backend: Backend,
+    settings: PhantomData<Settings>,
+}
+
+impl<Backend: RecoveryBackend, Settings> RecoveryOperator<Backend, Settings>
+where
+    Backend: RecoveryBackend + FromSettings + Clone + Debug,
+    Backend::State: Clone + Debug,
+    Settings: Clone + Debug,
+{
+    fn new(recovery_backend: Backend) -> Self {
+        Self {
+            recovery_backend,
+            settings: Default::default(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<Backend, Settings> StateOperator for RecoveryOperator<Backend, Settings>
+where
+    Settings: Send + Debug + Clone,
+    Backend: RecoveryBackend + FromSettings<Settings = Settings> + Debug + Clone + Send,
+    Backend::State: Serialize
+        + DeserializeOwned
+        + FromSettings<Settings = Settings>
+        + Send
+        + Default
+        + Debug
+        + Clone,
+{
+    type StateInput = RecoverableState<Backend, Settings>;
+
+    fn from_settings(settings: <Self::StateInput as ServiceState>::Settings) -> Self {
+        let recovery_backend = Backend::from_settings(&settings);
+        Self::new(recovery_backend)
+    }
+
+    async fn run(&mut self, state: Self::StateInput) {
+        let save_result = self
+            .recovery_backend
+            .save_state(&state.state)
+            .map_err(MyError::from);
+        if let Err(error) = save_result {
+            error!("{}", error);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use async_trait::async_trait;
+    use overwatch_derive::Services;
+    use overwatch_rs::overwatch::OverwatchRunner;
+    use overwatch_rs::services::handle::{ServiceHandle, ServiceStateHandle};
+    use overwatch_rs::services::relay::RelayMessage;
+    use overwatch_rs::services::{ServiceCore, ServiceData, ServiceId};
+    use overwatch_rs::DynError;
+    use std::env::temp_dir;
+    use std::path::PathBuf;
+
+    #[derive(Debug, Clone)]
+    struct FileBackend<State> {
+        directory: PathBuf,
+        state: PhantomData<State>,
+    }
```

Comment:
This can probably be promoted as part of the utils module itself. It would need some tweaks. The generic is not really necessary here I think. It can be part of the usage methods as long as the type implements `Deserialize`.
For an idea of a fn signature: `fn load<T: DeserializeOwned>(&self) -> T`

Reply:
I thought so as well but I didn't split anything into files yet as I thought of this as a proposal.